plugins/com.google.cloud.tools.eclipse.appengine.validation/xslt/upgradeRuntime.xsl

```diff
@@ -20,6 +20,16 @@
     </xsl:copy>
   </xsl:template>
 
+  <!-- add missing runtime element -->
+  <xsl:template match="appengine:appengine-web-app">
+    <xsl:copy>
+      <xsl:apply-templates select="node()|@*"/>
+      <xsl:if test="not(appengine:runtime)">  <runtime>java8</runtime><xsl:text>
+</xsl:text></xsl:if>
+    </xsl:copy>
+  </xsl:template>
+
+  <!-- upgrade existing runtime element -->
   <xsl:template match="appengine:runtime"><runtime>java8</runtime></xsl:template>
```

Comment:
Give two-space indentation before `<runtime>` like above?

Reply:
no, the indentation here is off to try to get the correct indentation in the generated output. That is some though not all whitespace in XSLT is significant.